core/src/main/java/com/linecorp/armeria/server/annotation/BeanRequestConverterFunction.java

```diff
@@ -458,77 +405,56 @@ protected String getParamStrValue(final RequestConvertContext context) {
-                            String.format("Annotation should not be used on parameter '%1$s' of '%2$s'.",
-                                          parameter.getName(), getMemberFullName(element))
-                    );
-                }
+        @Nullable
+        private static ParamArrayValueRetriever from(final Executable element) {
+            if (hasParamOrHeader(element)) {
+                return fromAnnotatedMethodOrConstructor(element);
+            } else {
+                return fromAnnotatedParameters(element);
             }
         }
 
-        private static ParamArrayValueRetriever buildFrom(final Executable element) {
-            final Annotation annotation = ParamValueRetriever.getParamsAnnotation(element);
+        private static ParamArrayValueRetriever fromAnnotatedMethodOrConstructor(Executable element) {
             final Parameter[] elementParams = element.getParameters();
-
-            if (annotation != null) {
-                // if there is annotation on this method/constructor
-
-                // there should be only one parameter
-                if (elementParams.length != 1) {
-                    throw new IllegalArgumentException(
-                            String.format(
-                                    "There should be only 1 parameter for '%1$s'.",
-                                    getMemberFullName(element))
-                    );
-                }
-
-                // parameter should have no annotation
-                checkNoParamAnnotations(element, elementParams);
-
-                final Class<?> type = elementParams[0].getType();
-                final ParamValueRetriever retriever = ParamValueRetriever.buildFrom(type, annotation);
-
-                return of(retriever);
-            }
-
-            // build retrievers by annotations on each parameters
-            final List<ParamValueRetriever> retrieverList = Arrays.stream(elementParams)
-                                                                  .map(ParamValueRetriever::buildFrom)
-                                                                  .filter(Objects::nonNull)
-                                                                  .collect(Collectors.toList());
-            if (retrieverList.isEmpty()) {
-                // there is no annotation on params of this method/constructor
+            checkArgument(elementParams.length == 1,
+                          "the number of parameters of %s: %s (expected: 1)",
+                          getMemberFullName(element), elementParams.length);
+            checkArgument(Arrays.stream(elementParams)
+                                .noneMatch(AnnotatedHttpServiceParamUtil::hasParamOrHeader),
+                          "no @%s and @%s annotations are allowed in the parameter of %s",
+                          Param.class.getSimpleName(), Header.class.getSimpleName(),
+                          getMemberFullName(element));
```

Comment:
Better using traditional `if` blocks because otherwise `Class.getSimpleName()` and `getMemberFullName()` will be evaluated even on successful path.